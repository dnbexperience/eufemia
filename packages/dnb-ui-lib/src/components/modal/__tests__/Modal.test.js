/**
 * Component Test
 *
 */

import React from 'react'
import {
  mount,
  fakeProps,
  toJson,
  axeComponent,
  loadScss
} from '../../../core/jest/jestSetup'

import Component from '../Modal'

const props = fakeProps(require.resolve('../Modal'), {
  all: true,
  optional: true
})
props.title = 'modal_title'
props.id = 'modal_id'
props.content_id = 'modal_content_id'
props.modal_content = 'unique_modal_content'
props.close_title = 'close_title'
props.direct_dom_return = true
props.no_animation = true
props.preventSetTriggerRef = true // we set preventSetTriggerRef to true, cause jest gives us an error

describe('Modal component', () => {
  const Comp = mount(<Component {...props} />)
  Comp.setState({
    modalActive: true
  })
  it('have to match snapshot', () => {
    expect(toJson(Comp)).toMatchSnapshot()
  })
  it('has to have the correct title', () => {
    expect(Comp.find('h1').text()).toBe(props.title)
  })
  it('has no trigger button once we set trigger_hidden to true', () => {
    Comp.setProps({
      trigger_hidden: true
    })
    expect(Comp.find('button.dnb-modal__trigger').exists()).toBe(false)
    Comp.setProps({
      trigger_hidden: false
    })
  })
  it('has a disabled trigger button once we set trigger_disabled to true', () => {
    Comp.setProps({
      trigger_disabled: true
    })
    expect(
      Comp.find('button.dnb-modal__trigger')
        .instance()
        .hasAttribute('disabled')
    ).toBe(true)
  })
  it('has working open event and close event if "Esc" key gets pressed', async () => {
    const on_close = jest.fn()
    const on_open = jest.fn()
    const Comp = mount(
      <Component {...props} on_close={on_close} on_open={on_open} />
    )
    Comp.find('button').simulate('click')

    await wait(15) // wait for the event to be called

    Comp.find('div.dnb-modal__content__inner').simulate('keyDown', {
      key: 'Esc',
      keyCode: 27
    })

    await wait(15) // wait for the event to be called

    expect(on_open).toHaveBeenCalled()
    expect(on_close).toHaveBeenCalled()
  })
  it('has working open event and close event on changing the "open_state"', async () => {
    const on_close = jest.fn()
    const on_open = jest.fn()
    const Comp = mount(
      <Component {...props} on_close={on_close} on_open={on_open} />
    )
    Comp.setProps({ open_state: 'opened' })
    await wait(10) // wait for the render to be finished
    expect(on_open).toHaveBeenCalled()

    Comp.setProps({ open_state: 'closed' })
    await wait(10) // wait for the render to be finished
    expect(on_close).toHaveBeenCalled()
  })
  it('runs expected side effects', async () => {
    const Comp = mount(<Component {...props} />)
    const elem = Comp.find('button')

    // open modal
    elem.simulate('click')

    await wait(10) // wait for the render to be finished

    // const body = document.querySelector('[data-dnb-modal-active]')
    expect(document.body.nodeName).toBe('BODY')
    expect(document.body.style.overflow).toBe('hidden')
    expect(document.body.style.height).toBe('auto')
    // expect(document.documentElement.style.height).toBe('auto')
    expect(document.body.getAttribute('data-dnb-modal-active')).toBe(
      'true'
    )

    // close modal
    elem.simulate('click')
    await wait(10) // wait for the render to be finished

    expect(document.body.style.position).not.toBe('hidden')
  })
  it('has expected open and close states', () => {
    const Comp = mount(<Component {...props} />)
    Comp.setProps({ open_state: 'opened' })

    expect(Comp.state().modalActive).toBe(true)

    Comp.setProps({ open_state: 'closed' })

    expect(Comp.state().modalActive).toBe(false)
  })
  it('has an opened modal if open_state is set to "opened"', () => {
    const Comp = mount(<Component {...props} />)
    Comp.setProps({
      open_state: 'opened'
    })
    expect(Comp.exists('div.dnb-modal__content')).toBe(true)
    Comp.setProps({
      open_state: 'closed'
    })
    expect(Comp.exists('div.dnb-modal__content')).toBe(false)
  })
  it('has to have the correct aria-describedby', () => {
    expect(
      Comp.find('[aria-describedby]').props()['aria-describedby']
    ).toBe(props.id)
  })
  it('has to have the correct role on aria-modal', () => {
    expect(Comp.find('[aria-modal]').props().role).toBe('dialog')
  })
  it('has to have a close button', () => {
    expect(
      Comp.find(`button[aria-label="${props.close_title}"]`).props()[
        'aria-label'
      ]
    ).toBe(props.close_title)
  })
  it('has to have no icon', () => {
    const Comp1 = mount(<Component trigger_text="Open Modal" />)
    expect(Comp1.find(`.dnb-icon`).exists()).toBe(false)
    const Comp2 = mount(
      <Component
        trigger_text="Open Modal"
        trigger_variant="tertiary"
        trigger_icon={null}
      />
    )
    expect(Comp2.find(`.dnb-icon`).exists()).toBe(false)
  })
  it('has to have an icon', () => {
    const Comp1 = mount(
      <Component trigger_text="Open Modal" trigger_variant="tertiary" />
    )
    expect(Comp1.find(`.dnb-icon`).exists()).toBe(true)
    const Comp2 = mount(
      <Component trigger_text="Open Modal" trigger_icon="add" />
    )
    expect(Comp2.find(`.dnb-icon`).exists()).toBe(true)
  })
  it('should validate with ARIA rules as a dialog', async () => {
    expect(await axeComponent(Comp)).toHaveNoViolations()
  })
})

describe('Modal scss', () => {
  it('have to match snapshot', () => {
    const scss = loadScss(require.resolve('../style/dnb-modal.scss'))
    expect(scss).toMatchSnapshot()
  })
})

const wait = (t) => new Promise((r) => setTimeout(r, t))
