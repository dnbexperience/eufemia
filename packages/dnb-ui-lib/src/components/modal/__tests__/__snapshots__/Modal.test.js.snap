// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component have to match snapshot 1`] = `
<Modal
  0={
    Object {
      "displayName": "Modal",
      "props": Object {
        "children": "children",
        "class": "class",
        "className": "className",
        "close_modal": [Function],
        "close_title": "close_title",
        "content_id": "content_id",
        "direct_dom_return": "direct_dom_return",
        "disabled": "disabled",
        "fullscreen": "fullscreen",
        "hide_close_button": "hide_close_button",
        "id": "id",
        "labelled_by": "labelled_by",
        "modal_content": "modal_content",
        "on_close": [Function],
        "on_close_prevent": [Function],
        "on_open": [Function],
        "open_modal": [Function],
        "open_state": "'opened'",
        "preventSetTriggerRef": true,
        "prevent_close": "prevent_close",
        "prevent_core_style": "prevent_core_style",
        "title": "title",
        "trigger_class": "trigger_class",
        "trigger_disabled": "trigger_disabled",
        "trigger_hidden": "trigger_hidden",
        "trigger_icon": "trigger_icon",
        "trigger_icon_position": "trigger_icon_position",
        "trigger_text": "trigger_text",
        "trigger_title": "trigger_title",
        "trigger_variant": "custom type not supported, please set the correct value for trigger_variant prop",
      },
    }
  }
  1={
    Object {
      "displayName": "ModalRoot",
      "props": Object {
        "children": "children",
        "direct_dom_return": "direct_dom_return",
      },
    }
  }
  2={
    Object {
      "displayName": "ModalContent",
      "props": Object {
        "children": "children",
        "class": "class",
        "className": "className",
        "closeModal": [Function],
        "close_title": "close_title",
        "content_id": "content_id",
        "fullscreen": "fullscreen",
        "hide_close_button": "hide_close_button",
        "labelled_by": "labelled_by",
        "max_width": "max_width",
        "min_width": "min_width",
        "modal_content": "modal_content",
        "prevent_close": "prevent_close",
        "prevent_core_style": "prevent_core_style",
        "title": "title",
      },
    }
  }
  3={
    Object {
      "displayName": "CloseButton",
      "props": Object {
        "className": "className",
        "on_click": [Function],
        "title": "title",
      },
    }
  }
  class={null}
  className={null}
  close_modal={null}
  close_title="close_title"
  content_id="modal_content_id"
  direct_dom_return={true}
  disabled={null}
  fullscreen={false}
  hide_close_button={false}
  id="modal_id"
  labelled_by={null}
  modal_content="unique_modal_content"
  on_close={null}
  on_close_prevent={null}
  on_open={null}
  open_modal={null}
  open_state={null}
  preventSetTriggerRef={true}
  prevent_close={false}
  prevent_core_style={false}
  title="modal_title"
  trigger_class={null}
  trigger_disabled={null}
  trigger_hidden={false}
  trigger_icon="question"
  trigger_icon_position="left"
  trigger_text={null}
  trigger_title={null}
  trigger_variant="secondary"
>
  <div
    className="dnb-modal"
  >
    <Button
      bounding={false}
      class={null}
      className="dnb-modal__trigger"
      custom_element={null}
      custom_method={null}
      disabled={false}
      href={null}
      icon="question"
      icon_position="left"
      icon_size={null}
      id="modal_id"
      innerRef={null}
      onClick={null}
      on_click={[Function]}
      size={null}
      text={null}
      title="modal_title"
      type="button"
      variant="secondary"
    >
      <button
        className="dnb-button dnb-button--secondary dnb-modal__trigger dnb-button--icon-position-left dnb-button--has-icon dnb-button--size-medium"
        id="modal_id"
        onClick={[Function]}
        onMouseOut={[Function]}
        title="modal_title"
        type="button"
      >
        <Content
          bounding={false}
          class={null}
          className="dnb-modal__trigger"
          content={null}
          custom_element={null}
          custom_method={null}
          disabled={false}
          href={null}
          icon="question"
          icon_position="left"
          icon_size={null}
          id="modal_id"
          innerRef={null}
          isIconOnly={true}
          onClick={null}
          on_click={[Function]}
          size={null}
          text={null}
          title="modal_title"
          type="button"
          variant="secondary"
        >
          â€Œ
          <IconPrimary
            alt={null}
            aria-hidden={true}
            attributes={null}
            border={null}
            class={null}
            className="dnb-button__icon"
            color={null}
            height={null}
            icon="question"
            key="button-icon"
            modifier={null}
            size={null}
            title={null}
            width={null}
          >
            <span
              alt="question"
              aria-hidden={true}
              className="dnb-icon dnb-icon--default dnb-button__icon"
              role="img"
            >
              <question>
                <svg
                  fill="none"
                  height={16}
                  viewBox="0 0 16 16"
                  width={16}
                >
                  <path
                    clipRule="evenodd"
                    d="M4.352 5.1c0-2.05 1.55-3.6 3.6-3.6 2.052 0 3.601 1.55 3.601 3.6 0 2.051-1.55 3.6-3.6 3.6a.75.75 0 00-.75.75v2.666a.75.75 0 001.5 0V10.15c2.506-.348 4.35-2.425 4.35-5.049 0-2.879-2.22-5.1-5.1-5.1-2.88 0-5.1 2.221-5.1 5.1a.75.75 0 001.5 0zm3.6 10.898c.659 0 1.098-.439 1.098-1.097 0-.658-.44-1.097-1.097-1.097-.658 0-1.097.44-1.097 1.097 0 .658.439 1.097 1.097 1.097z"
                    fill="#000"
                    fillRule="evenodd"
                  />
                </svg>
              </question>
            </span>
          </IconPrimary>
        </Content>
      </button>
    </Button>
    <ModalRoot
      0={
        Object {
          "displayName": "Modal",
          "props": Object {
            "children": "children",
            "class": "class",
            "className": "className",
            "close_modal": [Function],
            "close_title": "close_title",
            "content_id": "content_id",
            "direct_dom_return": "direct_dom_return",
            "disabled": "disabled",
            "fullscreen": "fullscreen",
            "hide_close_button": "hide_close_button",
            "id": "id",
            "labelled_by": "labelled_by",
            "modal_content": "modal_content",
            "on_close": [Function],
            "on_close_prevent": [Function],
            "on_open": [Function],
            "open_modal": [Function],
            "open_state": "'opened'",
            "preventSetTriggerRef": true,
            "prevent_close": "prevent_close",
            "prevent_core_style": "prevent_core_style",
            "title": "title",
            "trigger_class": "trigger_class",
            "trigger_disabled": "trigger_disabled",
            "trigger_hidden": "trigger_hidden",
            "trigger_icon": "trigger_icon",
            "trigger_icon_position": "trigger_icon_position",
            "trigger_text": "trigger_text",
            "trigger_title": "trigger_title",
            "trigger_variant": "custom type not supported, please set the correct value for trigger_variant prop",
          },
        }
      }
      1={
        Object {
          "displayName": "ModalRoot",
          "props": Object {
            "children": "children",
            "direct_dom_return": "direct_dom_return",
          },
        }
      }
      2={
        Object {
          "displayName": "ModalContent",
          "props": Object {
            "children": "children",
            "class": "class",
            "className": "className",
            "closeModal": [Function],
            "close_title": "close_title",
            "content_id": "content_id",
            "fullscreen": "fullscreen",
            "hide_close_button": "hide_close_button",
            "labelled_by": "labelled_by",
            "max_width": "max_width",
            "min_width": "min_width",
            "modal_content": "modal_content",
            "prevent_close": "prevent_close",
            "prevent_core_style": "prevent_core_style",
            "title": "title",
          },
        }
      }
      3={
        Object {
          "displayName": "CloseButton",
          "props": Object {
            "className": "className",
            "on_click": [Function],
            "title": "title",
          },
        }
      }
      class={null}
      className={null}
      closeModal={[Function]}
      close_modal={null}
      close_title="close_title"
      content_id="modal_content_id"
      direct_dom_return={true}
      fullscreen={false}
      hide_close_button={false}
      labelled_by="modal_id"
      modal_content="unique_modal_content"
      on_close={null}
      on_close_prevent={null}
      on_open={null}
      open_modal={null}
      prevent_close={false}
      prevent_core_style={false}
      title="modal_title"
      toggleOpenClose={[Function]}
    >
      <ModalContent
        0={
          Object {
            "displayName": "Modal",
            "props": Object {
              "children": "children",
              "class": "class",
              "className": "className",
              "close_modal": [Function],
              "close_title": "close_title",
              "content_id": "content_id",
              "direct_dom_return": "direct_dom_return",
              "disabled": "disabled",
              "fullscreen": "fullscreen",
              "hide_close_button": "hide_close_button",
              "id": "id",
              "labelled_by": "labelled_by",
              "modal_content": "modal_content",
              "on_close": [Function],
              "on_close_prevent": [Function],
              "on_open": [Function],
              "open_modal": [Function],
              "open_state": "'opened'",
              "preventSetTriggerRef": true,
              "prevent_close": "prevent_close",
              "prevent_core_style": "prevent_core_style",
              "title": "title",
              "trigger_class": "trigger_class",
              "trigger_disabled": "trigger_disabled",
              "trigger_hidden": "trigger_hidden",
              "trigger_icon": "trigger_icon",
              "trigger_icon_position": "trigger_icon_position",
              "trigger_text": "trigger_text",
              "trigger_title": "trigger_title",
              "trigger_variant": "custom type not supported, please set the correct value for trigger_variant prop",
            },
          }
        }
        1={
          Object {
            "displayName": "ModalRoot",
            "props": Object {
              "children": "children",
              "direct_dom_return": "direct_dom_return",
            },
          }
        }
        2={
          Object {
            "displayName": "ModalContent",
            "props": Object {
              "children": "children",
              "class": "class",
              "className": "className",
              "closeModal": [Function],
              "close_title": "close_title",
              "content_id": "content_id",
              "fullscreen": "fullscreen",
              "hide_close_button": "hide_close_button",
              "labelled_by": "labelled_by",
              "max_width": "max_width",
              "min_width": "min_width",
              "modal_content": "modal_content",
              "prevent_close": "prevent_close",
              "prevent_core_style": "prevent_core_style",
              "title": "title",
            },
          }
        }
        3={
          Object {
            "displayName": "CloseButton",
            "props": Object {
              "className": "className",
              "on_click": [Function],
              "title": "title",
            },
          }
        }
        class={null}
        className={null}
        closeModal={[Function]}
        close_modal={null}
        close_title="close_title"
        content_id="modal_content_id"
        fullscreen={false}
        hide_close_button={false}
        labelled_by="modal_id"
        max_width={null}
        min_width={null}
        modal_content="unique_modal_content"
        on_close={null}
        on_close_prevent={null}
        on_open={null}
        open_modal={null}
        prevent_close={false}
        prevent_core_style={false}
        title="modal_title"
        toggleOpenClose={[Function]}
      >
        <div
          aria-describedby="modal_id"
          aria-labelledby="modal_id"
          aria-modal="true"
          className="dnb-modal__content"
          onClick={[Function]}
          role="dialog"
        >
          <div
            className="dnb-modal__content__inner dnb-no-focus dnb-core-style"
            id="modal_content_id"
            onClick={[Function]}
            onKeyDown={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
          >
            <h1
              className="dnb-modal__title dnb-h2"
            >
              modal_title
            </h1>
            <CloseButton
              className={null}
              on_click={[Function]}
              title="close_title"
            >
              <Button
                aria-label="close_title"
                bounding={false}
                class={null}
                className="dnb-modal__close-button"
                custom_element={null}
                custom_method={null}
                disabled={null}
                href={null}
                icon="close"
                icon_position={null}
                icon_size="medium"
                id={null}
                innerRef={null}
                onClick={null}
                on_click={[Function]}
                size="medium"
                text={null}
                title={null}
                type="button"
                variant="secondary"
              >
                <button
                  aria-label="close_title"
                  className="dnb-button dnb-button--secondary dnb-modal__close-button dnb-button--icon-position-right dnb-button--has-icon dnb-button--icon-size-medium dnb-button--size-medium"
                  onClick={[Function]}
                  onMouseOut={[Function]}
                  type="button"
                >
                  <Content
                    aria-label="close_title"
                    bounding={false}
                    class={null}
                    className="dnb-modal__close-button"
                    content={null}
                    custom_element={null}
                    custom_method={null}
                    disabled={null}
                    href={null}
                    icon="close"
                    icon_position={null}
                    icon_size="medium"
                    id={null}
                    innerRef={null}
                    isIconOnly={true}
                    onClick={null}
                    on_click={[Function]}
                    size="medium"
                    text={null}
                    title={null}
                    type="button"
                    variant="secondary"
                  >
                    â€Œ
                    <IconPrimary
                      alt={null}
                      aria-hidden={false}
                      attributes={null}
                      border={null}
                      class={null}
                      className="dnb-button__icon"
                      color={null}
                      height={null}
                      icon="close"
                      key="button-icon"
                      modifier={null}
                      size="medium"
                      title={null}
                      width={null}
                    >
                      <span
                        alt="close"
                        aria-hidden={false}
                        aria-label="close"
                        className="dnb-icon dnb-icon--medium dnb-button__icon"
                        role="img"
                      >
                        <close_medium>
                          <svg
                            fill="none"
                            height={24}
                            viewBox="0 0 24 24"
                            width={24}
                          >
                            <path
                              clipRule="evenodd"
                              d="M4.78 3.914a.75.75 0 00-1.06 1.061l7.22 7.22-7.22 7.22a.75.75 0 001.06 1.06l7.22-7.22 7.22 7.22a.75.75 0 001.06-1.06l-7.22-7.22 7.22-7.22a.75.75 0 00-1.06-1.06L12 11.134l-7.22-7.22z"
                              fill="#000"
                              fillRule="evenodd"
                            />
                          </svg>
                        </close_medium>
                      </span>
                    </IconPrimary>
                  </Content>
                </button>
              </Button>
            </CloseButton>
            <div
              className="dnb-modal__wrapper"
            >
              unique_modal_content
            </div>
          </div>
        </div>
        <span
          aria-hidden="true"
          className="dnb-modal__overlay"
        />
      </ModalContent>
    </ModalRoot>
  </div>
</Modal>
`;

exports[`Modal scss have to match snapshot 1`] = `
"/*
* DNB Modal
*
*/
/**
 * This file is only used to make components independent
 * so that they can get imported individually, without the core styles
 *
 */
/*
 * Utilities
 */
/*
 * Scopes
 *
 */
/*
 * Document Reset
 *
 */
/*
* DNB Button
*
*/
/*
* DNB icon
*
*/
/*
* Icon component
*
*/
.dnb-icon {
  display: inline-block;
  font-size: 1rem;
  line-height: 1rem;
  color: inherit;
  width: 1em;
  height: 1em; }
  .dnb-icon img,
  .dnb-icon svg {
    width: inherit;
    height: inherit;
    shape-rendering: geometricPrecision;
    vertical-align: top; }
  .dnb-icon svg[width='100%'] {
    width: inherit; }
  .dnb-icon svg[height='100%'] {
    height: inherit; }
  .dnb-icon svg:not([fill]),
  .dnb-icon svg [fill] {
    fill: currentColor; }
  .dnb-icon svg [stroke] {
    stroke: currentColor; }
  .dnb-icon--small {
    font-size: 0.75rem; }
  .dnb-icon--default {
    font-size: 1rem; }
  .dnb-icon--medium {
    font-size: 1.5rem; }
  .dnb-icon--large {
    font-size: 2rem; }
  .dnb-icon--x-large {
    font-size: 2.5rem; }
  .dnb-icon--xx-large {
    font-size: 3rem; }
  .dnb-icon--custom-size {
    width: auto;
    height: auto; }
  .dnb-icon--auto {
    font-size: 1em; }
  .dnb-icon--auto > .dnb-icon--wrapper {
    display: inline-flex;
    align-items: center;
    justify-content: center; }
  h1 .dnb-icon,
  h2 .dnb-icon,
  h3 .dnb-icon,
  h4 .dnb-icon,
  h5 .dnb-icon,
  h6 .dnb-icon {
    vertical-align: middle; }
  @media screen and (-ms-high-contrast: none) {
    .dnb-icon {
      flex: none; } }

.dnb-button {
  font-family: var(--font-family-default);
  font-weight: var(--font-weight-default);
  font-size: var(--font-size-small);
  font-style: normal;
  line-height: var(--line-height-basis);
  color: var(--color-black-80, #333);
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  /**
  * 1. Remove repeating backgrounds in all browsers (opinionated).
  * 2. Add border box sizing in all browsers (opinionated).
  */
  /**
  * 1. Add text decoration inheritance in all browsers (opinionated).
  * 2. Add vertical alignment inheritance in all browsers (opinionated).
  */
  margin: 0;
  padding: 0; }
  .dnb-button *,
  .dnb-button ::before,
  .dnb-button ::after {
    background-repeat: no-repeat;
    /* 1 */
    box-sizing: border-box;
    /* 2 */ }
  .dnb-button ::before,
  .dnb-button ::after {
    text-decoration: inherit;
    /* 1 */
    vertical-align: inherit;
    /* 2 */ }

/*
* Button component
*
*/
:root {
  --button-font-size: var(--font-size-basis);
  --button-font-size-small: var(--font-size-small);
  --button-width: 2.5rem;
  --button-height: 2.5rem;
  --button-width--small: 1.5rem;
  --button-height--small: 1.5rem;
  --button-width--medium: 2rem;
  --button-height--medium: 2rem;
  --button-width--large: 3rem;
  --button-height--large: 3rem;
  --button-icon-size: 1rem;
  --button-border-width: 0.0625rem;
  --button-border-width--hover: 0.1875rem;
  --button-border-radius: calc(var(--button-height) / 2);
  --button-border-radius--small: calc(var(--button-height--small) / 2);
  --button-border-radius--medium: calc(var(--button-height--medium) / 2);
  --button-border-radius--large: calc(var(--button-height--large) / 2); }

.dnb-button {
  user-select: none;
  -webkit-user-select: none;
  cursor: pointer;
  white-space: nowrap;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  width: var(--button-width);
  height: var(--button-height);
  padding: 0;
  border: var(--button-border-width) solid transparent;
  border-radius: var(--button-border-radius);
  text-decoration: none;
  font-size: var(--font-size-small);
  line-height: var(--button-height--small); }
  .dnb-button__icon.dnb-icon svg:not([width]):not([height]) {
    width: var(--button-icon-size);
    height: var(--button-icon-size); }
  .dnb-button__text {
    font-size: var(--button-font-size);
    line-height: var(--line-height-x-small--em);
    color: inherit;
    background-color: inherit; }
  .dnb-button__bounding {
    position: absolute;
    background-color: transparent;
    width: var(--button-width);
    height: var(--button-height);
    border-radius: var(--button-border-radius);
    transform: scale(1.2); }
  .dnb-button--size-small {
    width: var(--button-width--small);
    height: var(--button-height--small);
    font-size: var(--button-font-size-small);
    line-height: var(--button-height--small);
    border-radius: var(--button-border-radius--small); }
  .dnb-button--has-text.dnb-button--size-small {
    padding-left: 1rem;
    padding-right: 1rem; }
  .dnb-button--has-text.dnb-button--icon-position-left.dnb-button--size-small {
    padding-left: 0.5rem; }
  .dnb-button--has-text.dnb-button--icon-position-right.dnb-button--size-small {
    padding-right: 0.5rem; }
  .dnb-button--size-medium {
    width: var(--button-width--medium);
    height: var(--button-height--medium);
    line-height: var(--button-height--medium);
    border-radius: var(--button-border-radius--medium); }
  .dnb-button--has-text.dnb-button--size-medium {
    padding-left: 1rem;
    padding-right: 1rem; }
  .dnb-button--has-text.dnb-button--icon-position-left.dnb-button--size-medium {
    padding-left: 0.5rem; }
  .dnb-button--has-text.dnb-button--icon-position-right.dnb-button--size-medium {
    padding-right: 0.5rem; }
  .dnb-button--size-large {
    width: var(--button-width--large);
    height: var(--button-height--large);
    line-height: var(--button-height--large);
    border-radius: var(--button-border-radius--large); }
  .dnb-button--has-text.dnb-button--size-large {
    padding-left: 2rem;
    padding-right: 2rem; }
  .dnb-button--has-text.dnb-button--icon-position-left.dnb-button--size-large {
    padding-left: 1rem; }
  .dnb-button--has-text.dnb-button--icon-position-right.dnb-button--size-large {
    padding-right: 1rem; }
  .dnb-button--has-text {
    padding-left: 1.5rem;
    padding-right: 1.5rem;
    width: auto; }
  .dnb-button--has-text .dnb-button__icon {
    margin: 0 calc(var(--button-icon-size) / 2); }
  .dnb-button--has-text.dnb-button--icon-position-left {
    padding-left: 0.5rem; }
  .dnb-button--has-text.dnb-button--icon-position-right {
    padding-right: 0.5rem; }
  .dnb-button--has-text.dnb-button--has-icon .dnb-button__icon {
    order: 2; }
  .dnb-button--has-text.dnb-button--has-icon .dnb-button__text {
    order: 1; }
  .dnb-button--has-text.dnb-button--has-icon.dnb-button--icon-position-left .dnb-button__icon {
    order: 1; }
  .dnb-button--has-text.dnb-button--has-icon.dnb-button--icon-position-left .dnb-button__text {
    order: 2; }
  .dnb-button--reset {
    margin: 0;
    padding: 0;
    width: auto;
    height: auto;
    overflow: visible;
    border: none;
    border-radius: 0;
    background-color: transparent;
    appearance: none;
    box-shadow: none;
    color: inherit;
    font: inherit;
    text-align: inherit;
    line-height: inherit; }
    html:not([data-is-touch]) .dnb-button--reset:hover[disabled] {
      cursor: not-allowed; }
    html:not([data-is-touch]) .dnb-button--reset:hover:not([disabled]) {
      box-shadow: none;
      border: none; }
    .dnb-button--reset:not([disabled]):focus, .dnb-button--reset:not([disabled]):active {
      outline: none; }
      html[data-whatinput='keyboard'] .dnb-button--reset:not([disabled]):focus, html[data-whatinput='keyboard'] .dnb-button--reset:not([disabled]):active {
        box-shadow: 0 0 0 0.125rem var(--color-emerald-green);
        /* IE11 fix */
        border-color: transparent; }
    html[data-whatinput='mouse'] .dnb-button--reset:not([disabled]):focus,
    html[data-whatinput='mouse'] .dnb-button--reset:not([disabled]):active {
      box-shadow: none;
      color: inherit;
      border: none; }
  .dnb-button[type='button'], .dnb-button[type='reset'], .dnb-button[type='submit'] {
    appearance: none;
    -moz-appearance: none;
    -webkit-appearance: none; }
  .dnb-button[disabled] {
    cursor: not-allowed; }
  .dnb-form-row--vertical .dnb-form-row__content > .dnb-button {
    align-self: flex-start; }
  @media screen and (-ms-high-contrast: none) {
    .dnb-button {
      flex: none; }
      .dnb-button__icon, .dnb-button__text {
        transform: translateY(-0.0625rem) !important; } }

/* Firefox includes a hidden border which messes up button dimensions */
button.dnb-button::-moz-focus-inner {
  border: none; }

.dnb-modal {
  font-family: var(--font-family-default);
  font-weight: var(--font-weight-default);
  font-size: var(--font-size-small);
  font-style: normal;
  line-height: var(--line-height-basis);
  color: var(--color-black-80, #333);
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  /**
  * 1. Remove repeating backgrounds in all browsers (opinionated).
  * 2. Add border box sizing in all browsers (opinionated).
  */
  /**
  * 1. Add text decoration inheritance in all browsers (opinionated).
  * 2. Add vertical alignment inheritance in all browsers (opinionated).
  */
  margin: 0;
  padding: 0; }
  .dnb-modal *,
  .dnb-modal ::before,
  .dnb-modal ::after {
    background-repeat: no-repeat;
    /* 1 */
    box-sizing: border-box;
    /* 2 */ }
  .dnb-modal ::before,
  .dnb-modal ::after {
    text-decoration: inherit;
    /* 1 */
    vertical-align: inherit;
    /* 2 */ }

/*
 * Modal component
 *
 */
:root {
  --modal-z-index: 3000;
  --modal-spacing: 2rem;
  --modal-max-width: 60rem; }

[data-dnb-modal-active='true'] {
  user-select: none;
  -webkit-user-select: none; }

#dnb-modal-root {
  position: absolute;
  top: 0;
  left: 0; }

.dnb-modal {
  display: inline-flex; }
  .dnb-modal__content {
    position: fixed;
    top: 0;
    left: 0;
    z-index: calc(var(--modal-z-index) + 1);
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    height: 100%;
    user-select: none;
    -webkit-user-select: none; }
    .dnb-modal__content__inner {
      position: relative;
      padding: var(--modal-spacing) var(--modal-spacing) 0;
      width: auto;
      height: auto;
      max-width: var(--modal-max-width);
      max-height: calc( 100vh - var(--modal-spacing) * 2);
      overflow-x: hidden;
      overflow-y: auto;
      user-select: text;
      -webkit-user-select: text;
      border: none;
      background: var(--color-white); }
      @media screen and (max-width: 40em) {
        .dnb-modal__content__inner {
          padding: calc(var(--modal-spacing) / 2) var(--modal-spacing) 0; } }
      @media screen and (min-width: 40em) {
        .dnb-modal__content__inner {
          padding: var(--modal-spacing) var(--modal-spacing) calc(var(--modal-spacing) / 2);
          box-shadow: 0 1px 6px rgba(0, 0, 0, 0.16);
          border-radius: 0.5rem; } }
      @media screen and (max-width: 40em) {
        .dnb-modal__content__inner {
          width: 100vw;
          height: 100vh;
          max-width: 100%;
          max-height: 100%;
          padding: calc(var(--modal-spacing) / 2) calc(var(--modal-spacing) / 2) 0; } }
    .dnb-modal__content--fullscreen {
      align-items: flex-start;
      justify-content: flex-start; }
    .dnb-modal__content--fullscreen .dnb-modal__content__inner {
      width: 100vw;
      height: 100vh;
      max-width: 100%;
      max-height: 100%; }
  .dnb-modal__title.dnb-h2,
  .dnb-core-style .dnb-modal__title.dnb-h2:not([class*='dnb-space']) {
    margin: 0;
    padding: 0 2rem 0 0; }
  .dnb-modal__wrapper {
    position: relative;
    z-index: 1;
    margin-top: calc(var(--modal-spacing) - 0.5rem);
    margin-bottom: var(--modal-spacing); }
    @media screen and (max-width: 40em) {
      .dnb-modal__wrapper {
        margin-top: calc(var(--modal-spacing) / 2);
        margin-bottom: calc(var(--modal-spacing) / 2); } }
    .dnb-modal__wrapper .dnb-section::after {
      left: calc(1px - 1px - var(--modal-spacing)); }
  .dnb-modal__close-button {
    position: absolute;
    top: var(--modal-spacing);
    right: var(--modal-spacing);
    z-index: 1; }
    @media screen and (max-width: 40em) {
      .dnb-modal__close-button {
        top: calc(var(--modal-spacing) / 2);
        right: calc(var(--modal-spacing) / 2); } }
  .dnb-modal__overlay {
    position: fixed;
    z-index: var(--modal-z-index);
    left: 0;
    top: 0;
    width: 100%;
    height: 100%;
    background-color: rgba(255, 255, 255, 0.85); }
"
`;
